File: socket_client_txt_input.py
import socket
import sys
import json
import argparse
from sys import argv
import time

SERVER_IP = socket.gethostname()
PORT = 5000


def client_program(port, exchange, outPut):
    host = SERVER_IP  # as both code is running on same pc

    # client_socket = socket.socket()  # instantiate
    try:
        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except socket.error as e:
        print("Error creating socket: %s" % e)
        sys.exit(1)

    try:
        client_socket.connect((host, port))
    except socket.gaierror as e:
        print("Address-related error connecting to server: %s" % e)
        sys.exit(1)
    except socket.error as e:
        print("Connection error: %s" % e)
        sys.exit(1)

    client_socket.send(exchange.encode())  # send message
    data = None
    while not data:
        data = client_socket.recv(1024).decode()

    client_socket.send(outPut.encode())  # send message
    data = None
    while not data:
        data = client_socket.recv(1024).decode()

    print("\nType 'bye' to end \n")

    message = input(" what Symbol? ")  # take input

    while message.lower().strip() != "bye":
        try:
            client_socket.send(message.encode())  # send message
        except socket.error as e:
            print("Error sending data: %s" % e)
            sys.exit(1)

        try:
            data = client_socket.recv(1024).decode()  # receive response
            if outPut == "json":
                data = json.loads(data)  # data loaded
        except socket.error as e:
            print("Error receiving data: %s" % e)
            sys.exit(1)

        info = ""
        if outPut == "json":
            i = 0
            for k, v in data.items():
                if str(v).isspace() or v == "":
                    v = "EMPTY"
                adding = f"{k} =  {v}\n"
                info = info + adding
                i += 1
        else:
            info = data
        if not data:
            info = "Wrong Data Entered"
        print("The info: \n" + str(info) + "\n")  # show in terminal

        message = input(" Symbol again: ?\n ")  # again take input

    client_socket.close()  # close the connection


if __name__ == "__main__":
    # port = int(input("enter port number: "))
    parser = argparse.ArgumentParser(
        description="""This is a very basic client program using tcp"""
    )

    parser.add_argument(
        "Exchange_name",
        type=str,
        help="Either NASDAQ, CME ,OR NYSE",
        action="store",
        choices=["NASDAQ", "CME", "NYSE"],
    )

    parser.add_argument(
        "outputType",
        type=str,
        help="This is info goes in the updated table",
        action="store",
        default="txt",
        nargs="?",
        choices=["txt", "json"],
    )

    args = parser.parse_args(argv[1:])
    client_program(PORT, args.Exchange_name, args.outputType)

File: socket_client_txt_input.cpp
// g++ socket_client_txt_input.cpp -std=c++17 -o client && ./client CME txt

#include <iostream>
#include <string>
#include <cstring>
#include <map>
#include <sstream>
#include <arpa/inet.h>
#include <unistd.h>
#include "nlohmann/json.hpp"

#define SERVER_IP "127.0.0.1"
#define PORT 5000

using json = nlohmann::json;

void client_program(int port, const std::string &exchange, const std::string &outputType)
{
    int client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket == -1)
    {
        perror("Error creating socket");
        exit(1);
    }

    sockaddr_in server_addr{};
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);

    if (inet_pton(AF_INET, SERVER_IP, &(server_addr.sin_addr)) <= 0)
    {
        perror("Address-related error connecting to server");
        exit(1);
    }

    if (connect(client_socket, reinterpret_cast<struct sockaddr *>(&server_addr), sizeof(server_addr)) < 0)
    {
        perror("Connection error");
        exit(1);
    }

    if (send(client_socket, exchange.c_str(), exchange.size(), 0) == -1)
    {
        perror("Error sending data");
        exit(1);
    }

    char buffer[1024] = {0};
    if (recv(client_socket, buffer, 1024, 0) == -1)
    {
        perror("Error receiving data");
        exit(1);
    }
    std::string response(buffer);

    if (send(client_socket, outputType.c_str(), outputType.size(), 0) == -1)
    {
        perror("Error sending data");
        exit(1);
    }

    memset(buffer, 0, sizeof(buffer));
    if (recv(client_socket, buffer, 1024, 0) == -1)
    {
        perror("Error receiving data");
        exit(1);
    }
    std::string data(buffer);

    std::cout << "\nType 'bye' to end\n";

    std::string message;
    std::cout << "What Symbol? ";
    while (std::getline(std::cin, message))
    {
        if (message == "bye")
        {
            break;
        }

        if (send(client_socket, message.c_str(), message.size(), 0) == -1)
        {
            perror("Error sending data");
            exit(1);
        }

        memset(buffer, 0, sizeof(buffer));
        if (recv(client_socket, buffer, 1024, 0) == -1)
        {
            perror("Error receiving data");
            exit(1);
        }
        std::string response(buffer);

        std::string info;
        if (outputType == "json")
        {
            json jsonData;
            try
            {
                jsonData = json::parse(response);
                int i = 0;
                for (const auto &[key, value] : jsonData.items())
                {
                    std::string strValue = value.get<std::string>();
                    if (strValue.empty() || std::isspace(strValue[0]))
                    {
                        strValue = "EMPTY";
                    }
                    std::ostringstream oss;
                    oss << key << " = " << strValue << "\n";
                    info += oss.str();
                    i++;
                }
            }
            catch (const json::parse_error &e)
            {
                std::cerr << "Error parsing JSON: " << e.what() << std::endl;
            }
        }
        else
        {
            info = response;
        }

        if (info.empty())
        {
            info = "Wrong Data Entered";
        }
        std::cout << "The info:\n"
                  << info << std::endl;

        std::cout << "Symbol again? ";
    }

    close(client_socket);
}

int main(int argc, char *argv[])
{
    if (argc != 3)
    {
        std::cerr << "Invalid number of arguments.\n";
        std::cerr << "Usage: ./client <Exchange_name> <outputType>\n";
        std::cerr << "Exchange_name: Either NASDAQ, CME, or NYSE\n";
        std::cerr << "outputType: txt or json\n";
        return 1;
    }

    int port = PORT;
    std::string exchange = argv[1];
    std::string outputType = argv[2];

    client_program(port, exchange, outputType);

    return 0;
}
